server/engine/src/main/java/org/eclipse/lsp/cobol/implicitDialects/cics/utility/CICSSignonOptionsCheckUtility.java

```diff
@@ -0,0 +1,95 @@
+import org.eclipse.lsp.cobol.common.error.SyntaxError;
+import org.eclipse.lsp.cobol.implicitDialects.cics.CICSLexer;
+import org.eclipse.lsp.cobol.implicitDialects.cics.CICSParser;
+
+import java.util.*;
+
+import static org.eclipse.lsp.cobol.implicitDialects.cics.CICSParser.RULE_cics_signon;
+import static org.eclipse.lsp.cobol.implicitDialects.cics.CICSParser.RULE_cics_signon_body;
+
+/** Checks CICS SIGNON rules for required and invalid options */
+public class CICSSignonOptionsCheckUtility extends CICSOptionsCheckBaseUtility {
+
+    public static final int RULE_INDEX = RULE_cics_signon;
+
+    private static final Map<Integer, ErrorSeverity> DUPLICATE_CHECK_OPTIONS =
+        new HashMap<Integer, ErrorSeverity>() {
+            {
+                put(CICSLexer.USERID, ErrorSeverity.ERROR);
+                put(CICSLexer.CHANGETIME, ErrorSeverity.ERROR);
+                put(CICSLexer.DAYSLEFT, ErrorSeverity.ERROR);
+                put(CICSLexer.ESMREASON, ErrorSeverity.ERROR);
+                put(CICSLexer.ESMRESP, ErrorSeverity.ERROR);
+                put(CICSLexer.EXPIRYTIME, ErrorSeverity.ERROR);
+                put(CICSLexer.GROUPID, ErrorSeverity.ERROR);
+                put(CICSLexer.INVALIDCOUNT, ErrorSeverity.ERROR);
+                put(CICSLexer.LANGUAGECODE, ErrorSeverity.ERROR);
+                put(CICSLexer.NATLANG, ErrorSeverity.ERROR);
+                put(CICSLexer.LANGINUSE, ErrorSeverity.ERROR);
+                put(CICSLexer.LASTUSETIME, ErrorSeverity.ERROR);
+                put(CICSLexer.NATLANGINUSE, ErrorSeverity.ERROR);
+                put(CICSLexer.PASSWORD, ErrorSeverity.ERROR);
+                put(CICSLexer.NEWPASSWORD, ErrorSeverity.ERROR);
+                put(CICSLexer.PHRASE, ErrorSeverity.ERROR);
+                put(CICSLexer.PHRASELEN, ErrorSeverity.ERROR);
+                put(CICSLexer.NEWPHRASE, ErrorSeverity.ERROR);
+                put(CICSLexer.NEWPHRASELEN, ErrorSeverity.ERROR);
+                put(CICSLexer.OIDCARD, ErrorSeverity.ERROR);
+            }
+        };
+
+    public CICSSignonOptionsCheckUtility(DialectProcessingContext context, List<SyntaxError> errors) {
+        super(context, errors, DUPLICATE_CHECK_OPTIONS);
+    }
+
+    /**
+     * Entrypoint to check CICS SIGNON rules for required and invalid options
+     * @param ctx ParserRuleContext subclass containing options
+     * @param <E> A subclass of ParserRuleContext
+     */
+    public <E extends ParserRuleContext> void checkOptions(E ctx) {
+        if (ctx.getRuleIndex() == RULE_cics_signon_body) {
+            checkRule((CICSParser.Cics_signon_bodyContext) ctx);
+            checkDuplicates(ctx);
+        }
+    }
+
+    private void checkRule(CICSParser.Cics_signon_bodyContext ctx) {
+        checkHasMandatoryOptions(ctx.USERID(), ctx, "USERID");
+
+        if (!ctx.NEWPHRASE().isEmpty())
+            checkHasMandatoryOptions(ctx.NEWPHRASELEN(), ctx, "NEWPHRASELEN");
```

Comment:
This and associated `checkPrerequisiteIsMet` could be handled by `checkAllOptionsArePresentOrAbsent` now.